src/Http/Http.Api/src/MapActionEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Http.Api;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    /// <summary>
+    /// Provides extension methods for <see cref="IEndpointRouteBuilder"/> to define HTTP API endpoints.
+    /// </summary>
+    public static class MapActionEndpointRouteBuilderExtensions
+    {
+        /// <summary>
+        /// Adds a <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that matches the pattern specified via attributes.
+        /// </summary>
+        /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <param name="action">The delegate executed when the endpoint is matched.</param>
+        /// <returns>A <see cref="IEndpointConventionBuilder"/> that can be used to further customize the endpoint.</returns>
+        public static IEndpointConventionBuilder MapAction(
+            this IEndpointRouteBuilder endpoints,
+            Delegate action)
```

Comment:
Linker unfriendly 😄 .

cc @eerhardt

Reply:
What's linker unfriendly about the API? You can have APIs that take delegates, and the linker won't trim them away.

Reply:
We do a ton of reflection on the delegate that's passed in.

Reply:
The `Expression` stuff is probably going to get you into more trouble. I'm in the middle of annotating Linq.Expressions now, and it isn't pretty how much stuff is not trim compatible.

Reply:
We're also considering Roslyn source generators as an alternative which I assume is more linker-friendly. @davidfowl has already started experimenting with this at https://github.com/davidfowl/uController/blob/aa8bcb4b30764e42bd72d326cb2718a4d4eaf4a9/src/uController.SourceGenerator/CodeGenerator.cs.

Reply:
Yeah, source generators are usually better, if possible